File: AIiOS/AIiOS/Services/RuleEngine.swift
import Foundation

/// Evaluates a `RouterRuleSet` against a fact dictionary. First matching rule
/// (in declared order) wins; falls back to `ruleSet.defaultDecision` if none match.
struct RuleEngine {

    struct Outcome {
        let decision: String   // "local" | "cloud"
        let reason:   String
        let ruleId:   String?  // nil when falling back to the default decision
    }

    func decide(facts: [String: FactValue], ruleSet: RouterRuleSet) -> Outcome {
        for rule in ruleSet.rules {
            if evaluate(rule.conditions, facts: facts) {
                return Outcome(decision: rule.decision, reason: rule.reason, ruleId: rule.id)
            }
        }
        return Outcome(decision: ruleSet.defaultDecision, reason: "No rule matched — default.", ruleId: nil)
    }

    // MARK: - Condition evaluation

    func evaluate(_ condition: RuleCondition, facts: [String: FactValue]) -> Bool {
        switch condition {
        case .all(let conds): return conds.allSatisfy { evaluate($0, facts: facts) }
        case .any(let conds): return conds.contains   { evaluate($0, facts: facts) }
        case .leaf(let leaf): return evaluate(leaf, facts: facts)
        }
    }

    private func evaluate(_ leaf: LeafCondition, facts: [String: FactValue]) -> Bool {
        guard let factValue = facts[leaf.fact] else { return false }
        guard let op = RuleOperator(rawValue: leaf.operator) else { return false }

        switch op {
        case .equal:
            return factValue == leaf.value
        case .notEqual:
            return factValue != leaf.value
        case .greaterThan:
            guard let a = factValue.asDouble, let b = leaf.value.asDouble else { return false }
            return a > b
        case .greaterThanInclusive:
            guard let a = factValue.asDouble, let b = leaf.value.asDouble else { return false }
            return a >= b
        case .lessThan:
            guard let a = factValue.asDouble, let b = leaf.value.asDouble else { return false }
            return a < b
        case .lessThanInclusive:
            guard let a = factValue.asDouble, let b = leaf.value.asDouble else { return false }
            return a <= b
        case .contains:
            guard case .string(let haystack) = factValue,
                  case .string(let needle)   = leaf.value else { return false }
            return haystack.localizedCaseInsensitiveContains(needle)
        }
    }
}

File: AIAndroid/app/src/main/kotlin/ai/ecoinference/app/router/RuleEngine.kt
package ai.ecoinference.app.router

/** First matching rule (in declared order) wins; falls back to `defaultDecision`. */
data class RuleOutcome(
    val decision: String,   // "local" | "cloud"
    val reason: String,
    val ruleId: String?,    // null when falling back to the default decision
)

/** Evaluates a [RouterRuleSet] against a fact map. Mirrors iOS RuleEngine.swift. */
object RuleEngine {

    fun decide(facts: Map<String, FactValue>, ruleSet: RouterRuleSet): RuleOutcome {
        for (rule in ruleSet.rules) {
            if (evaluate(rule.conditions, facts)) {
                return RuleOutcome(rule.decision, rule.reason, rule.id)
            }
        }
        return RuleOutcome(ruleSet.defaultDecision, "No rule matched — default.", null)
    }

    fun evaluate(condition: RuleCondition, facts: Map<String, FactValue>): Boolean = when (condition) {
        is RuleCondition.AllOf -> condition.conditions.all { evaluate(it, facts) }
        is RuleCondition.AnyOf -> condition.conditions.any { evaluate(it, facts) }
        is RuleCondition.Leaf  -> evaluateLeaf(condition, facts)
    }

    private fun evaluateLeaf(leaf: RuleCondition.Leaf, facts: Map<String, FactValue>): Boolean {
        val factValue = facts[leaf.fact] ?: return false
        return when (leaf.operator) {
            "equal"    -> factValue == leaf.value
            "notEqual" -> factValue != leaf.value
            "greaterThan" -> compare(factValue, leaf.value) { a, b -> a > b }
            "greaterThanInclusive" -> compare(factValue, leaf.value) { a, b -> a >= b }
            "lessThan" -> compare(factValue, leaf.value) { a, b -> a < b }
            "lessThanInclusive" -> compare(factValue, leaf.value) { a, b -> a <= b }
            "contains" -> {
                val haystack = (factValue as? FactValue.StringValue)?.value
                val needle   = (leaf.value as? FactValue.StringValue)?.value
                haystack != null && needle != null && haystack.contains(needle, ignoreCase = true)
            }
            else -> false
        }
    }

    private inline fun compare(a: FactValue, b: FactValue, op: (Double, Double) -> Boolean): Boolean {
        val ad = a.asDouble ?: return false
        val bd = b.asDouble ?: return false
        return op(ad, bd)
    }
}
